lib/HTFeed/JobMetrics.pm

```diff
@@ -308,73 +219,123 @@ sub _valid_metric {
     get_logger()->warn("invalid metric name \"$metric\" at $filename:$line");
     # switch from warn to die to ensure no invalid metric names slip thru:
     # die ("invalid metric name \"$metric\" at $filename:$line");
+
     return 0;
 }
 
+# Labels must be sent as a hashref with simple key-value pairs.
+# Not going to impose any stricter validation at this point.
+# Keys and values will be lowercased.
+# Labels do NOT need to be predefined, unlike metrics.
+sub _valid_labels {
+    my $self    = shift;
+    my $labels  = shift;
+
+    my $reftype = ref $labels;
+    if ($reftype ne "HASH") {
+        my (undef, $filename, $line) = caller(1);
+        get_logger()->warn("invalid labels (must be HASH ref) at $filename:$line");
+
+        return {};
+    }
+
+    # No nested data here, delete on sight.
+    foreach my $k (keys %$labels) {
+        my $v = $labels->{$k};
+        if (ref $v) {
+            get_logger()->warn(
+                "No nested values allowed in labels. Deleting label $k => $v\n"
+            );
+            delete $labels->{$k};
+            next;
+        }
+    }
+
+    # lowercase the whole hashref, now that keys & vals are known to be fine
+    $labels = { map lc, %$labels };
+
+    return $labels;
+}
+
 sub _setup_metrics {
     my $self = shift;
     my $prom = $self->{prom};
 
     # Prefix all metric names with:
     my $pfx = "ingest";
 
-    # Each element x in this array gets turned into 3 prom declarations:
-    # x_seconds, x_bytes and x_items
-    # Names are based on the downcased class name of the stage.
+    # Each element in @measurable_stages gets combined with
+    # each element in @units, to produce e.g. pack_items etc.
+    # Names are based on the downcased class name of the stage,
+    # or method in which the measured event happens.
     my @measurable_stages = (
-	'collate',
-	'download',
-	'download_dropbox',
-	'download_google',
-	'download_ia',
-	'handle',
-	'imageremediate',
-	'mets',
-	'pack',
-	'sourcemets',
-	'unpack',
-	'verifymanifest',
-	'volumevalidator'
+        'collate',
+        'download',
+        'download_dropbox',
+        'download_google',
+        'download_ia',
+        'handle',
+        'imageremediate',
+        'encrypt',
```

Comment:
I think if we find ourselves having to change this much more in the future we may want to rethink enumerating these here. I'm not sure how to track that sort of "hypothetical issue"... maybe that's a good place for a note in the code?

Reply:
I wanted/needed some structure to keep myself consistent as I was implementing metric names and labels. Totally free-form metric names would be near-useless, and metric names with typoes would be frustrating. But I agree that this is not sustainable and should be scaled back as we settle into use. (For instance, my early insistence that metrics must start with "ingest_" is already limiting/silly).

I think ultimately we can be satisfied with `/^\w+(_\w+)*$/`.

Reply:
And in case I didn't express this already, the main reason for this is because Prometheus requires that metrics be declared before they can be used. Perhaps we break this out as `JobMetrics::Definitions` which would be mostly lists of metric names and a few helper methods ...
```perl
package HTFeed::JobMetrics::Definitions;
my @counter_metrics = qw( ... );
foreach my $metric (@counter_metrics) {
    $prom->declare($metric_name, type => "counter");
} 
```

Flatter, dumber, easier to add/edit.

Reply:
Added issue #143 for this.